### PR TITLE
added ObsidianReader data connector

### PR DESCRIPTION
**Pull request by hursh-desai**
Useful for people that would like to quickly turn their Obsidian vault into a GPT Index that they can query. Most important feature to me is that I decided to chunk the markdown files by the headers instead of by the more random chunking that is usually used.

### CODE REVIEW

File: gpt_index/__init__.py [hunk 1]
```diff
@@ -73,6 +73,7 @@
     TwitterTweetReader,
     WeaviateReader,
     WikipediaReader,
+    ObsidianReader,
```
Comment (simonManydata) [inline gpt_index/__init__.py:76]: minor note:
I think the imports are sorted in alphabetical order.

You can sort imports automatically in VsCode:
https://eshlox.net/2019/12/02/vscode-sort-python-imports-automatically

File: gpt_index/readers/file/markdown_parser.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
+
+
+from abc import abstractmethod
+from pathlib import Path
+from typing import Dict, List, Optional, Union
+import re
+
+class MarkdownParser:
```
Comment (jerryjliu) [inline gpt_index/readers/file/markdown_parser.py:8]: i'd inherit from BaseParser

File: gpt_index/readers/file/markdown_parser.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
+
```
Comment (jerryjliu) [inline gpt_index/readers/file/markdown_parser.py:1]: nit: add docstring to top of file

Also bigger note, now that you've defined this parser, you can use this parser as part of SimpleDirectoryReader. Take a look at DEFAULT_FILE_EXTRACTOR in gpt_index.readers.file.base, you can add this as a parser to parse '.md' files.

File: gpt_index/readers/file/markdown_parser.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        return self._parser_config is not None
+
+    @property
+    def parser_config(self) -> Dict:
+        """Check if parser config is set."""
+        if self._parser_config is None:
+            raise ValueError("Parser config not set.")
+        return self._parser_config
+
+    @abstractmethod
```
Comment (jerryjliu) [inline gpt_index/readers/file/markdown_parser.py:71]: not an abstract method

File: gpt_index/readers/file/markdown_parser.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    def _init_parser(self) -> Dict:
+        """Initialize the parser with the config."""
+        parser_config = {'remove_hyperlinks': True, 'remove_images': True}
```
Comment (jerryjliu) [inline gpt_index/readers/file/markdown_parser.py:74]: note: you don't need to put this in _init_parser (this is really meant for loading "big" stuff like downloading a model from huggingface), you can just put these params as part of the `__init__`. Here you can just return an empty dictionary

File: gpt_index/readers/obsidian.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+from pathlib import Path
+from typing import Any, List
+import os
+
+from langchain.docstore.document import Document as LCDocument
+
+from gpt_index.readers.schema.base import Document
+from gpt_index.readers.file.markdown_parser import MarkdownParser
+
+class ObsidianReader:
```
Comment (jerryjliu) [inline gpt_index/readers/obsidian.py:12]: inherit from BaseReader

**Review by jerryjliu — COMMENTED** [review]
thanks for doing this @hursh-desai! this is huge 🥳 

left a few comments. also let me know if you need help linting/formatting

**Comment by jerryjliu** [comment]
@hursh-desai have you tested this? i could also test it on my end but if you've done it it'll save me the time :) 

also if you have tested it and have screenshots of an example usage, send it along! i'll tag it in the twitter post

**Comment by hursh-desai** [comment]
I've tested it for my own simple use case:
image
I have also included this notebook in the examples/data_connectors directory as ObsidianReaderDemo.ipynb.

I'm not sure how to fix 3 out of the 4 linting errors left, but there is one that I should point out is intentional.

Now that I am inheriting from BaseParser it is complaining that I am returning a dictionary from the `parse_file()` method, instead of a list like what the BaseParser class expects. This is intentional because I would like to keep the header that the text corresponds to. I'm hoping in the future I can include the header as metadata when indexing the documents.

**Comment by jerryjliu** [comment]
> I've tested it for my own simple use case: image I have also included this notebook in the examples/data_connectors directory as ObsidianReaderDemo.ipynb.
> 
> I'm not sure how to fix 3 out of the 4 linting errors left, but there is one that I should point out is intentional.
> 
> Now that I am inheriting from BaseParser it is complaining that I am returning a dictionary from the `parse_file()` method, instead of a list like what the BaseParser class expects. This is intentional because I would like to keep the header that the text corresponds to. I'm hoping in the future I can include the header as metadata when indexing the documents.



> I've tested it for my own simple use case: image I have also included this notebook in the examples/data_connectors directory as ObsidianReaderDemo.ipynb.
> 
> I'm not sure how to fix 3 out of the 4 linting errors left, but there is one that I should point out is intentional.
> 
> Now that I am inheriting from BaseParser it is complaining that I am returning a dictionary from the `parse_file()` method, instead of a list like what the BaseParser class expects. This is intentional because I would like to keep the header that the text corresponds to. I'm hoping in the future I can include the header as metadata when indexing the documents.

thanks for the image! 

no worries, i can take a look at this as well

**Comment by jerryjliu** [comment]
heads up @hursh-desai i did a few changes;
- cleaned up linting issues
- made your markdown dict a list of tuples instead of a dictionary of keys/values (it seems like preserving ordering would be nice no?) 
- removed some redundant overriding functions in the parser

**Comment by hursh-desai** [comment]
no problem. all of your changes made sense to me (probably should've used tuples from the start).
this is my first pr for an oss library so im honestly just excited to see my changes get merged lol.

**Review by jerryjliu — APPROVED** [review]

**Pull request merged**